internal/connection/connection.go

```diff
@@ -56,34 +103,110 @@ func NewConnection(ctx context.Context, endpoint, nodeID, driverName, namespace,
 	} else {
 		opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
 	}
-	cc, err := grpc.NewClient(endpoint, opts...)
+	cc, err := grpc.NewClient(c.endpoint, opts...)
 	if err != nil {
-		return nil, err
+		return err
+	}
+	c.Client = cc
+	c.connected = true
+
+	// Fetch the caps
+	// Downside is we need to do this on every re-connect
+	if err := c.fetchCapabilities(context.Background()); err != nil {
+		if e := c.Client.Close(); e == nil {
+			c.connected = false
+		}
+
+		return err
+	}
+
+	// Start a goroutine to close the connection after idle timeout
+	// But first, expire any existing timers
+	if c.cancelIdle != nil {
+		c.cancelIdle()
 	}
+	idleCtx, cFunc := context.WithCancel(context.Background())
+	c.cancelIdle = cFunc
+	go c.startIdleTimer(idleCtx)
+
+	return nil
+}
+
+// startIdleTimer starts a ticker with an interval of 30 seconds
+// At each tick, it checks if the connection has been idle for
+// more than `IDLE_TIMEOUT`. If so, the connection is closed.
+func (c *Connection) startIdleTimer(ctx context.Context) {
+	ticker := time.NewTicker(time.Second * 30)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.Lock()
+
+			lastAccess := time.Unix(0, c.lastAccessTime.Load())
+			isIdle := time.Since(lastAccess) > IDLE_TIMEOUT
+			if isIdle && c.connected {
+				// It's okay if there's an error in tearing down the connection
+				// It will be reused by subsequent requests, see Connect() for details
+				if err := c.Client.Close(); err == nil {
+					c.connected = false
+				}
+			}
+			c.Unlock()
+
+		case <-ctx.Done():
+			// Timer was cancelled
+			return
+		}
+	}
+}
+
+// NewConnection establishes connection with sidecar, fetches capability and returns Connection object
+// filled with required information.
+func NewConnection(ctx context.Context, endpoint, nodeID, driverName, namespace, podName string, enableAuth bool) (*Connection, error) {
 
 	conn := &Connection{
-		Client:     cc,
 		Namespace:  namespace,
 		Name:       podName,
 		NodeID:     nodeID,
 		DriverName: driverName,
 		Timeout:    time.Minute,
+
+		endpoint:   endpoint,
+		driverName: driverName,
+		podName:    podName,
+		enableAuth: enableAuth,
 	}
 
-	err = conn.fetchCapabilities(ctx)
-	if err != nil {
+	if err := conn.Connect(); err != nil {
 		return nil, err
 	}
 
 	return conn, nil
 }
 
+// Close tears down the gRPC connection and terminates the goroutines
+// monitoring the idle timeout by calling cancelIdle()
 func (c *Connection) Close() error {
-	if c.Client == nil {
-		return nil
+	c.Lock()
+	defer c.Unlock()
+
+	if c.cancelIdle != nil {
+		// Cancel the context as well
+		c.cancelIdle()
```

Comment:
set `c.cancelIdle = nil` after calling it?

Reply:
Done